### PR TITLE
Use aligned_alloc as we do in CoreNEURON too

### PR DESCRIPTION
**Pull request by ohm314**
also updated header includes

### CODE REVIEW

**Review by alexsavulescu — APPROVED** [review]
:rocket: 
nrn too?

**Review by alkino — APPROVED** [review]

**Comment by ohm314** [comment]
Noticed a performance degradation in our nightly benchmarks, so let's not merge yet

**Comment by ohm314** [comment]
decided to drop this since perf is impacted negatively.

**Comment by pramodk** [comment]
@ohm314 - what was last verdict on this part?

File: src/codegen/codegen_c_visitor.cpp [hunk 1]
```diff
@@ -1248,8 +1248,12 @@ void CodegenCVisitor::print_memory_allocation_routine() const {
     printer->add_newline(2);
     auto args = "size_t num, size_t size, size_t alignment = 16";
```
Comment (1uc) [inline src/codegen/codegen_c_visitor.cpp:1249]: I think AVX requires `32`; and AVX-512 `64` byte alignment.

File: src/codegen/codegen_c_visitor.cpp [hunk 2]
```diff
@@ -1248,8 +1248,12 @@ void CodegenCVisitor::print_memory_allocation_routine() const {
...
     printer->fmt_start_block("static inline void* mem_alloc({})", args);
-    printer->add_line("void* ptr;");
-    printer->add_line("posix_memalign(&ptr, alignment, num*size);");
+    printer->add_line("size_t fill = 0;");
+    printer->start_block("if (num % alignment != 0)");
+    printer->add_line("size_t multiple = num / alignment;");
+    printer->add_line("fill = alignment * (multiple + 1) - num;");
```
Comment (1uc) [inline src/codegen/codegen_c_visitor.cpp:1254]: I think the formula following formula for `ceil(num / alignment)` could come in handy:
```
(num*size + alignment - 1) / alignment) * alignment
```
(allows you to remove the `if` condition).

File: src/codegen/codegen_c_visitor.cpp [hunk 3]
```diff
@@ -1248,8 +1248,12 @@ void CodegenCVisitor::print_memory_allocation_routine() const {
...
+    printer->end_block(1);
+    printer->add_line("void* ptr = aligned_alloc(alignment, num*size + fill);");
```
Comment (1uc) [inline src/codegen/codegen_c_visitor.cpp:1256]: This looks strange when compared to `memset(ptr, 0, size)`.

**Comment by 1uc** [comment]
@ohm314 do you want to resolve the conflicts and then review/merge; or do we close this?

File: src/codegen/codegen_coreneuron_cpp_visitor.cpp [hunk 4]
```diff
@@ -330,11 +330,12 @@ bool CodegenCoreneuronCppVisitor::optimize_ion_variable_copies() const {
...
 void CodegenCoreneuronCppVisitor::print_memory_allocation_routine() const {
     printer->add_newline(2);
-    auto args = "size_t num, size_t size, size_t alignment = 16";
+    auto args = "size_t num, size_t size, size_t alignment = 32";
     printer->fmt_push_block("static inline void* mem_alloc({})", args);
-    printer->add_line("void* ptr;");
-    printer->add_line("posix_memalign(&ptr, alignment, num*size);");
-    printer->add_line("memset(ptr, 0, size);");
+    printer->add_line(
+        "size_t aligned_size = std::ceil(static_cast<float>(num*size) / alignment) * alignment");
```
Comment (ohm314) [inline src/codegen/codegen_coreneuron_cpp_visitor.cpp:336]: I guess that's what you had in mind to get rid of the extra `if` ?

File: src/codegen/codegen_coreneuron_cpp_visitor.cpp [hunk 5]
```diff
@@ -330,11 +330,12 @@ bool CodegenCoreneuronCppVisitor::optimize_ion_variable_copies() const {
 
 void CodegenCoreneuronCppVisitor::print_memory_allocation_routine() const {
     printer->add_newline(2);
-    auto args = "size_t num, size_t size, size_t alignment = 16";
+    auto args = "size_t num, size_t size, size_t alignment = 32";
```
[the remark this answers is not in the document]
Comment (pramodk) [reply src/codegen/codegen_coreneuron_cpp_visitor.cpp:333]: On coreneuron side, I see :

> coreneuron/utils/memory.h:#define NRN_SOA_BYTE_ALIGN (8 * sizeof(double))

**Review by 1uc — APPROVED** [review]

**Pull request merged**